### ISSUE TITLE
@b directive may delete ahb

### ISSUE THREAD

**Issue by Xilokar** [radare/radare2#9751]
### Work environment
Ubuntu x86 64
raw binary
arm/thumb
radare2 2.5.0-git 17697 @ linux-x86-64 git.2.4.0-178-ge0086d7d7
commit: e0086d7d701763c91f31c5875a47717138811210 build: 2018-03-20__14:02:38

### Expected behavior
@b should not delete any ahb set

### Actual behavior
@b may delete an ahb set

### Steps to reproduce the behavior 
```
r2 -aarm -b32 armthumb.bin
[0x00000000]> ahb 16 @0xc
[0x00000000]> ah
 0x0000000c - 0x0000000c => bits=16
[0x00000000]> pd 4 @b:32 @0x0000000c
            0x0000000c      0000           movs r0, r0
        ,=< 0x0000000e      a0e1           b 0x352
        |   0x00000010      0100           movs r1, r0
        |   0x00000012      00ebfeff       add.w pc, r0, lr, ror 31
[0x00000000]> ah
[0x00000000]> 
```
[armthumb.zip](https://github.com/radare/radare2/files/1837939/armthumb.zip)

**Comment by ret2libc** [comment]
Apart from the fact that I'm not getting exactly your output...
Here it seems there's some big problem with the way commands are parsed, more than with the `ahb/@b:` commands. (cc @radare)

If you do
```
[0x00000000]> pd 4 @ 0xc               
            0x0000000c      764c3615       ldrne r4, [r6, -0xc76]!             
            0x00000010      b1e31c00       ldrheq lr, [ip], -r1                
            0x00000014      00001c00       andseq r0, ip, r0                   
            0x00000018      00000c00       andeq r0, ip, r0          
```
Note the bytes (764c3615)
```
[0x00000000]> pd 4 @b:32 @ 0xc         
            0x0000000c      504b0304       streq r4, [r3], -0xb50              
            0x00000010      0a000000       andeq r0, r0, sl                    
            0x00000014      00004972       subvc r0, sb, 0                     
            0x00000018      764c3615       ldrne r4, [r6, -0xc76]!   
```
Now the bytes are different (the same is true if you use `@a` instead of `@b`)

**Comment by Xilokar** [comment]
(i think you are using the zip file instead of the armthumb.bin file inside the zip)

Indeed, as I was looking at bitness issues, I was just looking at the adresses, but you are right:

```
[0x00000000]> pd 8 @ 0x0
            0x00000000      0000a0e1       mov r0, r0
            0x00000004      010000eb       bl 0x10
        `-> 0x00000008      feffffea       b 8
            0x0000000c      01207047       ldrbmi r2, [r0, -r1]!
            0x00000010      00c09fe5       ldr ip, [pc]                ; [0x18:4]=13 ; 24
            0x00000014      1cff2fe1       bx ip
            0x00000018      0d000000       andeq r0, r0, sp
            0x0000001c      ffffffff       invalid
[0x00000000]> pd 4 @b:32 @ 0xc
            0x0000000c      0000a0e1       mov r0, r0
            0x00000010      010000eb       bl 0x1c
        `-> 0x00000014      feffffea       b 0x14
            0x00000018      01207047       ldrbmi r2, [r0, -r1]!
```
Using @b: clearly discard the temporarily seek address, but this seek address is nevertheless used in the displayed asm

**Comment by ret2libc** [comment]
Ops :P Yeah you are right, anyway the issue is there in any file.
It's not about `@b:` though, because you have the same problem when you use `@a:` too. I think it's actually a command parsing issue.

**Comment by Xilokar** [comment]
Now that the related issue raised by this has been fixed (by PR #9758), looking back at why ahb are deleted:
set_tmp_bits () do its stuff (searching for eventual hints to modify) before the temporary seek and the revert action (restoring hints) after.
IMHO, a better way would be to have set_tmp_bits set asm.bits and add a flag in RCore to prevent  r_core_seek_archbits() to change asm.bits if the flag if set.
This would also fix issue #9750 
A similar flag would also be needed for arch in the a: modifier.

Any comments ? (in particular regarding changing the RCore struct (abi ?))

**Comment by ret2libc** [comment]
@Xilokar When I modified it, I've done it as it was done elsewhere, but yes, I think your approach is better. I don't think adding a field to the RCore structure is a problem. :)
